### PR TITLE
Add missing constructors to ContextRoutes

### PR DESCRIPTION
**Pull request by hamnis**
Added some useful constructors to ContextRoutes that existed on HttpRoutes.

### CODE REVIEW

File: core/shared/src/main/scala/org/http4s/ContextRoutes.scala [hunk 1]
```diff
@@ -73,4 +75,50 @@ object ContextRoutes {
...
+    * is suspended in `F` to permit more efficient combination of
+    * routes via `SemigroupK`.
+    *
+    * @tparam F the effect of the [[ContextRoutes]]
+    * @tparam T the type of the context info in the [[ContextRequest]] accepted by the [[ContextRoutes]]
+    * @param routes the [[ContextRoutes]] to transform
+    * @return an [[ContextRoutes]] that wraps `run`
+    */
+  def toHttpRoutes[F[_]: Monad](routes: ContextRoutes[Unit, F]): HttpRoutes[F] =
+    Kleisli(req => Monad[OptionT[F, *]].unit >> routes.run(ContextRequest((), req)))
```
Comment (danicheg) [inline core/shared/src/main/scala/org/http4s/ContextRoutes.scala:123]: wdyt?
```suggestion
  def toHttpRoutes[T: Monoid, F[_]](routes: ContextRoutes[T, F]): HttpRoutes[F] =
    toHttpRoutes(Monoid[T].empty)(routes)

  def toHttpRoutes[T, F[_]](emptyContext: T)(routes: ContextRoutes[T, F]): HttpRoutes[F] =
    Kleisli(req => routes.run(ContextRequest(emptyContext, req)))
```

File: core/shared/src/main/scala/org/http4s/ContextRoutes.scala [hunk 2]
```diff
@@ -73,4 +76,63 @@ object ContextRoutes {
...
+    * @tparam F the base effect of the [[ContextRoutes]]
+    * @tparam T the type of the context info in the [[ContextRequest]] accepted by the [[ContextRoutes]]
+    * @param f  a function to apply to the [[ContextRequest]]
+    * @param fa the [[ContextRoutes]] to transform
+    * @return An [[ContextRoutes]] whose input is transformed by `f` before
+    *         being applied to `fa`
+    */
+  def local[T, F[_]: Monad](f: ContextRequest[F, T] => ContextRequest[F, T])(
+      fa: ContextRoutes[T, F]
+  ): ContextRoutes[T, F] = Kleisli(req => Monad[OptionT[F, *]].unit >> fa.run(f(req)))
```
Comment (danicheg) [inline core/shared/src/main/scala/org/http4s/ContextRoutes.scala:112]: Perhaps I'm missing something, but do we need here flatmapping over `F[Unit]`?

**Review by danicheg — APPROVED** [review]
Thanks for iterating over this! Despite my minor comment, this looks good to me

File: core/shared/src/main/scala/org/http4s/ContextRoutes.scala (hunk 2 above)
Comment (satorg) [reply core/shared/src/main/scala/org/http4s/ContextRoutes.scala:112]: I guess it is similar to `Http.local` – i.e. deferring without the `Defer` constraint per se.

Comment (danicheg) [reply core/shared/src/main/scala/org/http4s/ContextRoutes.scala:112]: Yeah, thank you. I rather meant, is this a thing actually?

Comment (danicheg) [reply core/shared/src/main/scala/org/http4s/ContextRoutes.scala:112]: Is it all about awareness of whether the `F` has eager semantics of computation?

**Pull request merged**